static/src/javascripts/projects/common/utils/steady-page.js

```diff
@@ -0,0 +1,187 @@
+/*
+ Utility: steady-insert.js
+ Insert components into web-pages without jumping them around by
+ moving the current scroll position by the same distance as the height of the
+ inserted element.
+ Note: Inserted elements must have a fixed height when rendered
+ (i.e: no changing the height by loading an image, lazy-loading content etc)
+ */
+
+define([
+    'common/utils/fastdom-promise',
+    'Promise',
+    'lodash/objects/assign',
+    'common/utils/queue',
+    'common/utils/config'
+], function (
+    fastdom,
+    Promise,
+    assign,
+    Queue,
+    config
+) {
+
+    var q = new Queue();
+    var running = false;
+    var promise;
+
+    /**
+     * Process the insertion operation:
+     *
+     *   1. Calculate the original height of the container
+     *   2. Apply the insertion functions for all inserted elements
+     *   3. Calculate the new height of the container
+     *   4. Adjust the scroll position to account for the new container height
+     *
+     * @param  {Number} state.scrollY The original scroll position
+     */
+    function go(state) {
+        running = true;
+
+        var batch = [];
+        var batchHeightsBeforeInsert;
+
+        promise = fastdom.read(function(){
+            while (!q.empty()) {
+                // Take the current queue items and add them to the batch array
+                batch.push(q.dequeue());
+            }
+
+            return batch;
+        })
+            .then(function(){
+                return getHeightOfAllContainers(batch);
+            })
+            .then(function(heightsBeforeIns) {
+                batchHeightsBeforeInsert = heightsBeforeIns || 0;
+                return insertElements(batch);
+            })
+            .then(function(){
+                return getHeightOfAllContainers(batch).then(function(newHeights) {
+                    return assign(state, {
+                        newHeight: newHeights - batchHeightsBeforeInsert
+                    });
+                });
+            })
+            .then(function(state){
+                if (q.empty()) {
+                    // If the queue is empty (no more elements need to be added to the page) we immediately scroll
+                    return state.newHeight + state.prevHeight + state.scrollY;
+                } else {
+                    // If there are elements waiting to be added to the page we take the previous container's heights
+                    // and recursively call the function so that we only scroll the page once the queue is empty -
+                    // this prevents excessive and jarring scrolling
+                    return go(assign(state, {
```

Comment:
I've got the feeling that this recursion is not tested in your specs. I might be wrong 😆